### PR TITLE
NSExpression initial implementation. With unit tests.

### PR DESCRIPTION
**Pull request by cacaodev**
This is the NSExpression implementation.
In *Apple Foundation Obj-C*, NSExpression  is a class cluster and instances are created via constructors (class methods). In *Apple Foundation Swift*, instances are created via regular initializers.
Following the last approach, I chose to implement NSExpression with closures because it is compact and simple. Each initializer defines 2 closures for evaluation and variables substitution.

Not implemented and incompatible with swift:
```swift
init(forFunction target: NSExpression, selectorName name: String, arguments parameters: [NSExpression]?)
class func expressionForAnyKey() -> NSExpression
```
Implemented but cannot be tested because dependent on another implementation:
```swift
init(forKeyPath:String) depends on NSObject valueForKeyPath(String)
init(forConditional...) and init(forSubquery...) depend on NSPredicate.
```
Most pre-defined function used in init(forFunction:arguments:) need an implementation.

Some methods have their signature changed compared to Apple Foundation in this PR:
The first one have a radar already opened : rdar://23937671
```swift
-expressionValueForObject(_:AnyObject?, context:NSMutableDictionary?) -> AnyObject 
-expressionValueForObject(_:AnyObject?, context:NSMutableDictionary?) -> AnyObject? 
``` 

```swift
init(forSubquery expression: NSExpression, usingIteratorVariable variable: String, predicate: AnyObject) to 
init(forSubquery expression: NSExpression, usingIteratorVariable variable: String, predicate: NSPredicate)

init(forAggregate subexpressions: [AnyObject]) to
init(forAggregate subexpressions: [NSExpression])

init(forFunction name: String, arguments parameters: [AnyObject]) to 
init(forFunction name: String, arguments parameters: [NSExpression])
```

The following method should be discussed IMO for change (not modified in this PR) because  NSExpression documentation clearly states that this method should throw if the function name is incorrect.
```swift
forFunction(_:String, arguments:[NSExpression] ) throws -> AnyObject? instead of
forFunction(_:String, arguments:[NSExpression] ) -> AnyObject?
```

TestNSExpression.swift have not been tested on Linux.

Future improvements: it is interesting to note that theoretically all expressions types except KeyPath and Function can have their evaluated value inferred from their initial values. Generics might help here.

### CODE REVIEW

File: Foundation/NSExpression.swift [hunk 1]
```diff
@@ -28,40 +28,152 @@ public enum NSExpressionType : UInt {
...
+        }
+
+        let evaluationBlock:EvaluateBlockType = {obj, bindings in
+            var result:AnyObject?
+            let args = parameters.expressionValueWithObject(obj, context: bindings)
+
+            do {
+                 result = try _invokeFunction(fn, with:args)
+            } catch let NSExpressionError.InvalidArgumentType(expected: type){
+                // TODO: Decide if we should throw instead. (Apple Foundation throws an Exception)
```
Comment (phausler) [inline Foundation/NSExpression.swift]: throw is reserved for error cases, exception cases are fatalError or precondition

File: Foundation/NSExpression.swift [hunk 2]
```diff
@@ -81,56 +194,358 @@ public class NSExpression : NSObject, NSSecureCoding, NSCopying {
...
+    public convenience init(forSubquery expression: NSExpression, usingIteratorVariable variable: String, predicate: NSPredicate) {
+        let evaluationBlock:EvaluateBlockType = {object, context in
+            guard let collection = expression.expressionValueWithObject(object, context: context) as? Array<AnyObject> else {
+                return nil
+            }
+
+            var context_dict:[String:AnyObject]
+
+            if let unwrapped = context {
+                context_dict = unwrapped as! Dictionary<String, AnyObject>
```
Comment (phausler) [inline Foundation/NSExpression.swift]: This fails to compile on linux:
```
Foundation/NSExpression.swift:276:42: warning: cast from 'NSMutableDictionary' to unrelated type 'Dictionary<String, AnyObject>' always fails
                context_dict = unwrapped as! Dictionary<String, AnyObject>
                               ~~~~~~~~~ ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: Foundation/NSExpression.swift [hunk 3]
```diff
@@ -81,56 +194,358 @@ public class NSExpression : NSObject, NSSecureCoding, NSCopying {
...
+            }
+
+            let result = collection.filter{predicate.evaluateWithObject($0, substitutionVariables:context_dict)}
+
+            return result.bridge()
+        }
+
+        let substitutionVariablesBlock:SubstituteBlockType = {bindings in
+            let exp = expression.expressionWithSubstitutionVariables(bindings)
+            let p = predicate.predicateWithSubstitutionVariables(bindings as! Dictionary<String,AnyObject>)
```
Comment (phausler) [inline Foundation/NSExpression.swift]:
```
Foundation/NSExpression.swift:291:75: warning: cast from 'NSMutableDictionary' to unrelated type 'Dictionary<String, AnyObject>' always fails
            let p = predicate.predicateWithSubstitutionVariables(bindings as! Dictionary<String,AnyObject>)
                                                                 ~~~~~~~~ ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
`

File: Foundation/NSExpression.swift [hunk 4]
```diff
@@ -81,56 +194,358 @@ public class NSExpression : NSObject, NSSecureCoding, NSCopying {
...
+    // TODO: Conditional Expression depends on NSPredicate implementation
+    public convenience init(forConditional predicate: NSPredicate, trueExpression: NSExpression, falseExpression: NSExpression) {
+        let evaluationBlock:EvaluateBlockType = {object, context in
+            let eval:Bool = predicate.evaluateWithObject(object, substitutionVariables: context as? Dictionary<String,AnyObject>)
+            let exp = eval ? trueExpression : falseExpression
+            return exp.expressionValueWithObject(object, context: context)
+        }
+
+        let substitutionVariablesBlock:SubstituteBlockType = {bindings in
+            let p = predicate.predicateWithSubstitutionVariables(bindings as! Dictionary<String,AnyObject>),
```
Comment (phausler) [inline Foundation/NSExpression.swift]:
```
Foundation/NSExpression.swift:333:75: warning: cast from 'NSMutableDictionary' to unrelated type 'Dictionary<String, AnyObject>' always fails
            let p = predicate.predicateWithSubstitutionVariables(bindings as! Dictionary<String,AnyObject>),
                                                                 ~~~~~~~~ ^   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by phausler** [comment]
The as casts need to be re-worked to use ._swiftObject and convert the keys accordingly. furthermore this has a compiler crash lurking in the current incantation (perhaps it is the `as` casts)

**Comment by parkera** [comment]
cc @amellifera

**Comment by parkera** [comment]
I spent a bunch of time trying to get this up to date with master and fixing the Linux problems but I ran out of time. I'm really sorry but I'll have to close this for now and ask to reopen once we can get the merge conflicts sorted out.

**Pull request closed without merging**